### PR TITLE
Feature/integration manager

### PR DESCRIPTION
**Pull request by ganfra**
Handle #48 issue

### CODE REVIEW

File: vector/src/main/res/layout/fragment_room_detail.xml [hunk 1]
```diff
@@ -108,15 +108,31 @@
         app:layout_constraintTop_toBottomOf="@id/syncStateView"
         tools:listitem="@layout/item_timeline_event_base" />
 
-    <im.vector.riotx.core.ui.views.JumpToReadMarkerView
-        android:id="@+id/jumpToReadMarkerView"
+
+    <FrameLayout
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_room_detail.xml:112]: This should be a Vertical LinearLayout, to be able to display both children

File: vector/src/main/res/layout/fragment_room_widget.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            android:src="@drawable/error" />
+
+        <TextView
+            android:id="@+id/widgetErrorText"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center"
+            android:layout_marginStart="@dimen/layout_horizontal_margin"
+            android:layout_marginEnd="@dimen/layout_horizontal_margin"
+            android:textColor="?android:textColorPrimary"
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_room_widget.xml:48]: Use ?riotx_text_primary

File: vector/src/main/res/xml/vector_settings_general.xml [hunk 3]
```diff
@@ -96,6 +96,25 @@
...
 
+    <im.vector.riotx.core.preference.VectorPreferenceCategory android:title="@string/settings_integrations">
+
+        <im.vector.riotx.core.preference.VectorPreference
+            android:focusable="false"
+            android:persistent="false"
+            android:summary="@string/settings_integrations_summary" />
+
+        <im.vector.riotx.core.preference.VectorSwitchPreference
+            android:key="SETTINGS_ALLOW_INTEGRATIONS_KEY"
```
Comment (bmarty) [inline vector/src/main/res/xml/vector_settings_general.xml:107]: Does this setting have to be persistent?

File: vector/src/main/res/layout/fragment_room_detail.xml (hunk 1 above)
Comment (ganfra) [reply vector/src/main/res/layout/fragment_room_detail.xml:112]: Nope, design by Rok want the jump to be on top of the widget

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/powerlevels/PowerLevelsHelper.kt [hunk 4]
```diff
@@ -44,15 +43,15 @@ class PowerLevelsHelper(private val powerLevelsContent: PowerLevelsContent) {
      * @param userId          the user id
      * @return true if the user can send this type of event
      */
-    fun isAllowedToSend(eventType: String, userId: String): Boolean {
-        return if (eventType.isNotEmpty() && userId.isNotEmpty()) {
+    fun isAllowedToSend(isState: Boolean, eventType: String?, userId: String): Boolean {
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/powerlevels/PowerLevelsHelper.kt:46]: I want to let the user to be able to set any kind of type, and we can't presume all state events. Maybe it's not defined in our sdk, so we have to fallback on something. The method is EventType.isStateEvent is not totally accurate

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/widgets/model/WidgetContent.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                }
+                id != null              -> {
+                    "$type $id"
+                }
+                else                    -> {
+                    "$type widget"
+                }
+            }
+        } else {
+            "Widget $id"
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/widgets/model/WidgetContent.kt:55]: I think widget shouldn't be translated actually. But this method is just duplicated from Legacy...

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/model/SessionRealmModule.kt [hunk 6]
```diff
@@ -55,6 +55,8 @@ import io.realm.annotations.RealmModule
             HomeServerCapabilitiesEntity::class,
             RoomMemberSummaryEntity::class,
             CurrentStateEventEntity::class,
-            UserAccountDataEntity::class
+            UserAccountDataEntity::class,
+            ScalarTokenEntity::class,
+            WellknownIntegrationManagerConfigEntity::class
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/model/SessionRealmModule.kt:60]: I think it is acceptable. The issue is when we perform clear cache we shouldn't just clear all tables, but select the ones we want to clear, as we would do with a sqlite db.

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/sync/model/accountdata/UserAccountDataWidgets.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+   },
+{
+      "..."
+   }
+}
+ */
+@JsonClass(generateAdapter = true)
+internal data class UserAccountDataWidgets(
+        @Json(name = "type") override val type: String = TYPE_WIDGETS,
+        @Json(name = "content") val content: Map<String, Event>
```
[the remark this answers is not in the document]
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/sync/model/accountdata/UserAccountDataWidgets.kt:49]: Yes it's a bit boring, but almost everything is event in matrix :D

File: matrix-sdk-android/src/main/res/values/integration_manager_config.xml [hunk 8]
```diff
@@ -1,11 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
 
-    <!-- Widget urls -->
-    <string name="integrations_ui_url" translatable="false">"https://scalar.vector.im/"</string>
-    <string name="integrations_rest_url" translatable="false">"https://scalar.vector.im/api"</string>
-    <string name="integrations_jitsi_widget_url" translatable="false">"https://scalar.vector.im/api/widgets/jitsi.html"</string>
-
     <string-array name="integrations_widgets_urls" translatable="false">
```
Comment (bmarty) [inline matrix-sdk-android/src/main/res/values/integration_manager_config.xml:4]: Is it used?

File: matrix-sdk-android/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -82,6 +82,13 @@
     <string name="notice_room_third_party_registered_invite">%1$s accepted the invitation for %2$s</string>
     <string name="notice_room_third_party_registered_invite_by_you">You accepted the invitation for %1$s</string>
 
+    <string name="notice_widget_added">%1$s widget added by %2$s</string>
+    <string name="notice_widget_added_by_you">%1$s widget added by you</string>
+    <string name="notice_widget_removed">%1$s widget removed by %2$s</string>
+    <string name="notice_widget_removed_by_you">%1$s widget removed by you</string>
+    <string name="notice_widget_modified">%1$s widget modified by %2$s</string>
+    <string name="notice_widget_modified_by_you">%1$s widget modified by you</string>
```
Comment (bmarty) [inline matrix-sdk-android/src/main/res/values/strings.xml:90]: Thanks for the update with "by_you" feature. For those wordings, are they copied from Riot-Web? I think "bob added widget X" will be more coherent with other notice string.

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/format/NoticeEventFormatter.kt [hunk 10]
```diff
@@ -88,10 +89,26 @@ class NoticeEventFormatter @Inject constructor(private val sessionHolder: Active
         val previousWidgetContent: WidgetContent? = event.prevContent.toModel()
         return if (widgetContent.isActive()) {
             val name = widgetContent.getHumanName()
-            sp.getString(R.string.event_formatter_widget_added, name, disambiguatedDisplayName)
+            if (previousWidgetContent?.isActive().orFalse()) {
+                if (event.isSentByCurrentUser()) {
+                    sp.getString(R.string.notice_widget_modified_by_you, name, disambiguatedDisplayName)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/format/NoticeEventFormatter.kt:94]: disambiguatedDisplayName is not needed here

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/format/NoticeEventFormatter.kt [hunk 11]
```diff
@@ -88,10 +89,26 @@ class NoticeEventFormatter @Inject constructor(private val sessionHolder: Active
...
+                } else {
+                    sp.getString(R.string.notice_widget_modified, name, disambiguatedDisplayName)
+                }
+            } else {
+                if (event.isSentByCurrentUser()) {
+                    sp.getString(R.string.notice_widget_added_by_you, name, disambiguatedDisplayName)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/format/NoticeEventFormatter.kt:100]: same remark

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/format/NoticeEventFormatter.kt [hunk 12]
```diff
@@ -88,10 +89,26 @@ class NoticeEventFormatter @Inject constructor(private val sessionHolder: Active
...
+                } else {
+                    sp.getString(R.string.notice_widget_added, name, disambiguatedDisplayName)
+                }
+            }
         } else {
-            val name = previousWidgetContent?.getHumanName() ?: sp.getString(R.string.room_widget_activity_title)
-            sp.getString(R.string.event_formatter_widget_removed, name, disambiguatedDisplayName)
+            val name = previousWidgetContent?.getHumanName()
+            if (event.isSentByCurrentUser()) {
+                sp.getString(R.string.notice_widget_removed_by_you, name, disambiguatedDisplayName)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/format/NoticeEventFormatter.kt:108]: same remark

File: vector/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -1108,8 +1108,9 @@
     <!-- Widget-->
     <string name="widget_no_power_to_manage">You need permission to manage widgets in this room</string>
     <string name="widget_creation_failure">Widget creation has failed</string>
-    <string name="event_formatter_widget_added">%1$s added by %2$s</string>
-    <string name="event_formatter_widget_removed">%1$s removed by %2$s</string>
+    <string name="event_formatter_widget_added">%1$s widget added by %2$s</string>
+    <string name="event_formatter_widget_removed">%1$s widget removed by %2$s</string>
+    <string name="event_formatter_widget_modified">%1$s widget modified by %2$s</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:1113]: Duplicated strings, already defined in the SDK?

File: vector/src/main/res/layout/fragment_room_detail.xml (hunk 1 above)
Comment (bmarty) [reply vector/src/main/res/layout/fragment_room_detail.xml:112]: 🤮

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/powerlevels/PowerLevelsHelper.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/powerlevels/PowerLevelsHelper.kt:46]: ok

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/model/SessionRealmModule.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/model/SessionRealmModule.kt:60]: ok

File: matrix-sdk-android/src/main/res/values/strings.xml (hunk 9 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/res/values/strings.xml:90]: Yes that's copied from riot web, not sure we should change it...

File: matrix-sdk-android/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -82,12 +82,12 @@
...
     <string name="notice_room_third_party_registered_invite_by_you">You accepted the invitation for %1$s</string>
 
-    <string name="notice_widget_added">%1$s widget added by %2$s</string>
-    <string name="notice_widget_added_by_you">%1$s widget added by you</string>
-    <string name="notice_widget_removed">%1$s widget removed by %2$s</string>
-    <string name="notice_widget_removed_by_you">%1$s widget removed by you</string>
-    <string name="notice_widget_modified">%1$s widget modified by %2$s</string>
-    <string name="notice_widget_modified_by_you">%1$s widget modified by you</string>
+    <string name="notice_widget_added">%1$s added %2$s widget</string>
+    <string name="notice_widget_added_by_you">you added %1$s widget</string>
```
Comment (bmarty) [inline matrix-sdk-android/src/main/res/values/strings.xml:86]: Start with upper case please (3 times)

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/widgets/DefaultWidgetURLFormatter.kt [hunk 15]
```diff
@@ -53,11 +52,10 @@ internal class DefaultWidgetURLFormatter @Inject constructor(private val integra
...
             currentConfig = preferredConfig
-            val defaultWhiteList = stringProvider.getStringArray(R.array.integrations_widgets_urls).asList()
-            whiteListedUrls = when (preferredConfig.kind) {
-                IntegrationManagerConfig.Kind.DEFAULT    -> defaultWhiteList
-                IntegrationManagerConfig.Kind.ACCOUNT    -> defaultWhiteList + preferredConfig.restUrl
-                IntegrationManagerConfig.Kind.HOMESERVER -> listOf(preferredConfig.restUrl)
+            whiteListedUrls = if (matrixConfiguration.integrationWidgetUrls.isEmpty()) {
+                listOf(preferredConfig.restUrl)
+            } else {
+                matrixConfiguration.integrationWidgetUrls
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/widgets/DefaultWidgetURLFormatter.kt:58]: This code is not equivalent (no use of `kind` anymore)? Maybe it's ok, just wanted to mention it

Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/widgets/DefaultWidgetURLFormatter.kt:58]: Make same logic than iOS and web

**Pull request merged**